File: source/RevitDevTool.Test/FaceVisualization.cs
using Autodesk.Revit.Attributes;
using Autodesk.Revit.UI.Selection;
using Nice3point.Revit.Toolkit.External;
using System.Diagnostics;

namespace RevitDevTool.Test;

[Transaction(TransactionMode.Manual)]
[UsedImplicitly]
public class FaceVisualization : ExternalCommand
{
    public override void Execute()
    {
        try
        {
            var faceRef = UiDocument.Selection.PickObject(ObjectType.Face, "Select Face");
            var face = Document.GetElement(faceRef)?.GetGeometryObjectFromReference(faceRef) as Face;
            Trace.Write(face);
        }
        catch (Exception e)
        {
            Trace.TraceError($"Error in FaceVisualization: {e.Message}");
        }
    }
}

[Transaction(TransactionMode.Manual)]
[UsedImplicitly]
public class FacesVisualization : ExternalCommand
{
    public override void Execute()
    {
        try
        {
            var faceRefs = UiDocument.Selection.PickObjects(ObjectType.Face, "Select Faces");
            var faces = new List<Face>();
            foreach (var faceRef in faceRefs)
            {
                var face = Document.GetElement(faceRef)?.GetGeometryObjectFromReference(faceRef) as Face;
                if (face != null)
                {
                    faces.Add(face);
                }
                else
                {
                    Trace.TraceWarning($"Face not found for reference: {faceRef}");
                }
            }

            Trace.Write(faces);
        }
        catch (Exception e)
        {
            Trace.TraceError($"Error in FacesVisualization: {e.Message}");
        }
    }
}

File: source/Samples/FSharpDemo/FaceVisualization.fs
namespace FSharpDemo

open System.Diagnostics
open Autodesk.Revit.Attributes
open Autodesk.Revit.DB
open Autodesk.Revit.UI.Selection
open JetBrains.Annotations
open Nice3point.Revit.Toolkit.External

[<Transaction(TransactionMode.Manual)>]
[<UsedImplicitly>]
type FaceVisualization() =
    inherit ExternalCommand()

    override this.Execute() =
        try
            let faceRef = this.UiDocument.Selection.PickObject(ObjectType.Face, "Select Face")
            let face = this.Document.GetElement(faceRef).GetGeometryObjectFromReference(faceRef) :?> Face
            Trace.Write(face)
        with
        | ex -> Trace.TraceError($"Error in FaceVisualization: {ex.Message}")

[<Transaction(TransactionMode.Manual)>]
[<UsedImplicitly>]
type FacesVisualization() =
    inherit ExternalCommand()

    override this.Execute() =
        try
            let faceRefs = this.UiDocument.Selection.PickObjects(ObjectType.Face, "Select Faces")
            let faces =
                [ for faceRef in faceRefs do
                    match this.Document.GetElement(faceRef).GetGeometryObjectFromReference(faceRef) with
                    | :? Face as face -> yield face
                    | _ -> Trace.TraceWarning($"Face not found for reference: {faceRef}") ]
            Trace.Write(faces)
        with
        | ex -> Trace.TraceError($"Error in FacesVisualization: {ex.Message}")

